### PR TITLE
fix for get_first_neighbor if there is only one particle

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This PR fixes an issue that was introduced in #3288 which causes the code to crash in the case where there is only one atom in the simulation cell. 

The distance tables function get_first_neighbor should only be called if there are actually neighbors, so an assertion is added in each of these functions. 

Also, in EinsplineSetBuilderESHDF.fft.cpp, now the get_first_neighbor is only called if there are multiple atoms in the simulation cell. 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
my workstation, gnu 9.2.0

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp [hunk 1]
```diff
@@ -182,7 +182,11 @@ bool EinsplineSetBuilder::ReadOrbitalInfo_ESHDF(bool skipChecks)
     const auto& ii_table = SourcePtcl->getDistTable(table_id);
     SourcePtcl->update(true);
     for (int i = 0; i < IonPos.size(); i++)
-      for (int j = 0; j < Super2Prim.size(); j++)
+    {
+      AtomicCentersInfo.non_overlapping_radius[i] = std::numeric_limits<RealType>::max();
+      //should only call get_first_neighbor to set non_overlapping_radius if there are more than one atom  in the cell
+      for (int j = 0; j < Super2Prim.size() && (Super2Prim.size() > 1); j++)
```
Comment (ye-luo) [inline src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp:188]: better to have `Super2Prim.size() > 1` condition standalone. and make it more explicitly visible.

File: src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp [hunk 2]
```diff
@@ -182,7 +182,11 @@ bool EinsplineSetBuilder::ReadOrbitalInfo_ESHDF(bool skipChecks)
...
+      {
```
Comment (ye-luo) [inline src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp:189]: No need of this pair {}

**Comment by prckent** [comment]
Thanks for spotting and fixing this Cody.

I am disappointed that this bug got in. Exactly what would be needed to hit this in a test? I think we need some 1 particle tests, e.g. 1 atom 1 electron case. @ye-luo Any thoughts on an existing test we could clone to avoid adding more wavefunction files?

**Comment by PDoakORNL** [comment]
Refactor until you can write a unit test. There appears to be no testing of this input parsing at all. You can generate the hdf5 file in the setup to the unit test it doesn't need to contain a real wavefunction.

**Comment by lshulen** [comment]
Strongly disagree.  The breakage should be fixed.  Then a unit test should be a high priority.  We should not leave the code in a broken state while we await perfection.

**Comment by prckent** [comment]
This is easy - we merge this fix. If Cody wants+is able to address Ye's minor comments, fine. But we can't leave the code broken, and the length of the ball of string/depth of can of worms could be massive here.

@camelto2 's choice on path forward.

**Comment by ye-luo** [comment]
We were relying an undefined behavior before #3288 and #3288 indeed adds assertion and this broken corner case got caught. it is unreasonable to ask very PR to fix previously owned debts. We just need to be realistically maximize cost effectiveness.

@prckent we have atomO test in pw2qmcpack workflow, It is not hitting this issue because of "n n n". We chose to put it in pw2qmcpack workflow in order to avoid storing the h5. I also found most of our solid tests has two atoms, we need to diversify them.  For testing purpose, we actually can add a cheap one atom system with low cutoff and test deterministically only. If we decide to add one case, @Hyeondeok-Shin can help @camelto2 on this. @prckent what do you think?

**Comment by PDoakORNL** [comment]
I just commented, I didn't request changes, I don't disagree with @lshulen and @prckent on the fix.

test_einset.cpp seems to encompass what we have for this so far although it is an integration test of SPOSet, SPOSetBuilderFactory, and the consequent subset of EinsplineSetBuilder. Whereas the test here should be straightforward and of EinsplineSetBuilder::ReadOrbitalInfo_ESHDF. When faced with a 1 particle ESHDF behave in the manner expected.  
Obviously there is a missing test for the no bugged case of >1 patricle ESHDF.

It's a bit of a challenge since the member function only mutates the internal state of ESB but that's part of the pain of having a testable code.

**Comment by PDoakORNL** [comment]
I'll disagree with @ye-luo a bit here. I think when you have a bug, especially one that will potentially recur and it can be caught in a unit test, there should be a unit test added. It deserves at least as much credit as fixing the bug after all if it had been written already there wouldn't have been one.

**Comment by prckent** [comment]
For a later PR: We can add a 1 atom case with low cutoff. There are already Li and H pseudopotentials in the repo that could be used for as a basis for this. We require both the statistical and deterministic versions. If the test added has one electron, the variance will be tiny, so a "short" statistical test will be accurate. It would be best to know + use the exact numerical result for this case. e.g. run HF in QE.

**Comment by ye-luo** [comment]
> For a later PR: We can add a 1 atom case with low cutoff. There are already Li and H pseudopotentials in the repo that could be used for as a basis for this. We require both the statistical and deterministic versions. If the test added has one electron, the variance will be tiny, so a "short" statistical test will be accurate. It would be best to know + use the exact numerical result for this case. e.g. run HF in QE.

I think both Li and H pseudo need high PW cutoff to have reasonable variance. Better to avoid them.

**Comment by ye-luo** [comment]
> I'll disagree with @ye-luo a bit here. I think when you have a bug, especially one that will potentially recur and it can be caught in a unit test, there should be a unit test added. It deserves at least as much credit as fixing the bug after all if it had been written already there wouldn't have been one.

I think once we got the simulation test with h5 provided, test_einset.cpp like unit test will be easy to add.

**Comment by camelto2** [comment]
> Since there has been some discussion, just to summarize the situation: @camelto2 is welcome to update as suggested by @ye-luo above, but we can merge this and work on expanding the tests or making other improvements later.

sorry for taking so long on this, I was out on Friday so missed all of this discussion. I'll go ahead and fix the minor issues from Ye and think about an effective test for a future PR

File: src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp (hunk 2 above)
Comment (camelto2) [reply src/QMCWaveFunctions/EinsplineSetBuilderESHDF.fft.cpp]: done

**Comment by Hyeondeok-Shin** [comment]
I can add deterministic test for a single atom case discussed in here. single Li atom will be good to make the test with low cut-off. I'll work on it.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Review by prckent — APPROVED** [review]
